oonib/testhelpers/http_helpers.py

```diff
@@ -168,7 +241,336 @@ def all(self, length, keyword):
+                'lock': defer.DeferredLock()
+            }
+            lock = self._cache_lifecycle[response_type][key_hash]['lock']
+            yield lock.acquire()
+            with open(cache_file, 'w+') as fw:
+                json.dump(value, fw)
+            lock.release()
+
+    @defer.inlineCallbacks
+    def expire(self, response_type, key_hash):
+        if response_type not in self._response_types:
+            raise Exception("Invalid response type")
+        lifecycle = self._cache_lifecycle[response_type][key_hash]
+        if lifecycle['expiration'].active():
+            lifecycle['expiration'].cancel()
+
+        yield lifecycle['lock'].acquire()
+        try:
+            os.remove(os.path.join(self._cache_dir, response_type, key_hash))
+        except OSError:
+            pass
+        lifecycle['lock'].release()
+        del self._cache_lifecycle[response_type][key_hash]
+
+    @defer.inlineCallbacks
+    def lookup(self, response_type, key):
+        if not self.enable_caching:
+            defer.returnValue(None)
+
+        key_hash = sha256(key).hexdigest()
+        cache_file = os.path.join(self._cache_dir, response_type, key_hash)
+
+        if key_hash not in self._cache_lifecycle[response_type]:
+            defer.returnValue(None)
+
+        lock = self._cache_lifecycle[response_type][key_hash]['lock']
+        expiration = \
+            self._cache_lifecycle[response_type][key_hash]['expiration']
+
+        yield lock.acquire()
+
+        if not os.path.exists(cache_file):
+            lock.release()
+            defer.returnValue(None)
+
+        with open(cache_file, 'r') as fh:
+            value = json.load(fh)
+
+        expiration.reset(self.expiration_time)
+        lock.release()
+        defer.returnValue(value)
+
+    @defer.inlineCallbacks
+    def http_request(self, url):
+        cached_value = yield self.lookup('http_request', url)
+        if cached_value is not None:
+            defer.returnValue(cached_value)
+
+        page_info = {
+            'body_length': -1,
+            'status_code': -1,
+            'headers': {},
+            'failure': None
+        }
+
+        agent = ContentDecoderAgent(FixedRedirectAgent(Agent(reactor)),
+                                    [('gzip', GzipDecoder)])
+        try:
+            retries = 0
+            while True:
+                try:
+                    response = yield agent.request('GET', url,
+                                                   TrueHeaders(REQUEST_HEADERS))
+                    headers = {}
+                    for name, value in response.headers.getAllRawHeaders():
+                        headers[name] = unicode(value[0], errors='ignore')
+                    body_length = -1
+                    try:
+                        body = yield readBody(response)
+                        body_length = len(body)
+                    except PartialDownloadError as pde:
```

Comment:
I mean that [the seventh -- in order of importance -- possible ways to indicate body length is](https://tools.ietf.org/html/rfc7230#section-3.3.3):

> 7.  Otherwise, this is a response message without a declared message
>      body length, so the message body length is determined by the
>      number of octets received prior to the server closing the
>      connection.

This way of indicating body length is not optimal, because obviously one cannot tell whether the server rightfully closed connection or whether, say, an error occurred forcing shutdown of the child serving the request. Yet, it is used in the wild in many cases.

So, my question was whether this Twisted error is used to indicate that specific use case of HTTP or whether Twisted had another way of reporting this case. Was asking this because, depending on the response, I am not sure to consider it correct to consider it good enough to use it.